gr-network/lib/tcp_source_impl.cc

```diff
@@ -0,0 +1,329 @@
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+
+#if defined(_WIN32) || defined(_WIN64)
+
+#define WINDOWS_SOCKETS
+
+#pragma commend(lib, "Ws2_32.lib")
+
+#include <Winsock2.h>
+#include <ws2tcpip.h>
+
+#else
+
+#include <arpa/inet.h>
+#include <asm-generic/socket.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#endif
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "tcp_source_impl.h"
+
+namespace gr {
+namespace network {
+
+typedef struct addrinfo addrinfo_t;
+
+tcp_source::sptr tcp_source::make(size_t itemsize,
+                                  size_t veclen,
+                                  const std::string& host,
+                                  const std::string& port,
+                                  int source_mode)
+{
+    return gnuradio::make_block_sptr<tcp_source_impl>(
+        itemsize, veclen, host, port, source_mode);
+}
+
+tcp_source_impl::tcp_source_impl(size_t item_size,
+                                 size_t vec_len,
+                                 const std::string& host,
+                                 const std::string& port,
+                                 int source_mode)
+    : d_connection_mode(source_mode), d_block_size(item_size * vec_len), d_veclen(vec_len)
+{
+#ifdef WINDOWS_SOCKETS
+
+    WSADATA wsaData;
+
+    int wsa_res = WSAStartup(MAKEWORD(2, 2), &wsaData);
+
+    if (wsa_res != 0) {
+        d_logger->error("WSAStartup failed with error code {}.", res);
+        throw std::runtime_error("Error while initializing Windows sockets.");
+    }
+
+    if (LOBYTE(wsaData.wVersion) != 2 || HIBYTE(wsaData.wVersion) != 2) {
+        d_logger->error("Winsock version 2.2 not available.");
+        WSACleanup();
+        throw std::runtime_error("Error while initializing Windows sockets.");
+    }
+
+#endif
+
+    struct addrinfo hints;
+    // very ugly but its either that or make an entire class for a single pointer that
+    // gets used a single time
+    std::unique_ptr<addrinfo_t, void (*)(addrinfo_t*)> host_info(new addrinfo_t,
+                                                                 freeaddrinfo);
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_flags = AI_PASSIVE;
+
+    // Get IP version, port, e.t.c in a workable format
+    int res;
+    addrinfo_t* tmp;
+    if (d_connection_mode == TCP_SOURCE_MODE_CLIENT) {
+        res = getaddrinfo(host.c_str(), port.c_str(), &hints, &tmp);
+    } else {
+        res = getaddrinfo(NULL, port.c_str(), &hints, &tmp);
+    }
+    host_info.reset(tmp);
+
+    if (res != 0) {
+#ifdef WINDOWS_SOCKETS
+        d_logger->erorr("Error while initializing: {}", WSAGetLastError());
+#else
+        d_logger->error("Error while initializing: {}", gai_strerror(res));
+#endif
+        throw std::runtime_error("TCP Source encountered an error while initializing");
+    }
+
+    bool failed = false;
```

Comment:
Probably easier to instead have `bool success = false`, and only set it just before `break`?